### Importing required libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, roc_curve, roc_auc_score,f1_score,auc


### Loading the data set

In [2]:
df=pd.read_csv("D:\\python_datascience\\data sets\\titanic\\train.csv")
df.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

### Checking shape of data 

In [3]:
df.shape

(891, 12)

### Checking data types of columns

In [4]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Age, Cabin and Embarked  has 177,687 and 2 missing values  

### Deleting the some columns which are not required for predictions

In [5]:
df.drop(columns=["PassengerId","Name","Cabin","Ticket"],inplace=True)

In [6]:
df

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female   NaN      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

### Checking the missing values 

In [7]:
df.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [8]:
## Handle Age columns with its mean value since it is of numeric type
m=df["Age"].mean()
print(m)
df["Age"].fillna(df["Age"].mean(),inplace=True)


29.69911764705882


In [9]:
## Handle Embarked column  missing values with its mode value since it is of object type
k=df["Embarked"].mode()
print(k)
df["Embarked"].fillna(k[0],inplace=True)

0    S
Name: Embarked, dtype: object


### checking missing values again

In [10]:
df.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

 now the data has no missing values 

#### label encoding of sex column as  :


In [11]:
df["Sex"].replace({"male":0,"female":1},inplace=True)


#### label encoding of Embarked column as :

In [12]:
print(df["Embarked"].unique())
df["Embarked"].replace({"S":0,"C":1,"Q":2},inplace=True)

['S' 'C' 'Q']


In [13]:
df.head(2)

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    0  22.0      1      0   7.2500         0
1         1       1    1  38.0      1      0  71.2833         1

In [14]:
x=df.drop(columns=["Survived"])
x.head(2) ## features

Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0       3    0  22.0      1      0   7.2500         0
1       1    1  38.0      1      0  71.2833         1

In [15]:
y=df["Survived"]
y.head(2) ## target

0    0
1    1
Name: Survived, dtype: int64

### Dividing the data set into train and test data as follows 

In [16]:
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=3) ## by default = 0.25 

### shape of train and test data 

In [17]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(668, 7)
(223, 7)
(668,)
(223,)


### Model fitting using random forest 

In [18]:
from sklearn.ensemble import RandomForestClassifier
clf=RandomForestClassifier()  ## taking default values  like bootstarp=True,critera=gini,n_estimators=100(100 trees or 100 individualmoddels),clss_weight = 0  means whether you want to give equal importance to both the class or make one as special class and many more


In [19]:
clf.fit(x_train,y_train)

RandomForestClassifier()

In [20]:
y_pred=clf.predict(x_test)
y_pred

array([0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1,
       0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0,
       1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 1], dtype=int64)

In [21]:
acc=accuracy_score(y_test,y_pred)
acc

0.7937219730941704

In [22]:
false_positive_rate,true_positive_rate,thresholds=roc_curve(y_test,y_pred)
roc_auc=auc(false_positive_rate,true_positive_rate)
roc_auc

0.7909947377355288

### hyper parameter tuning


In [24]:
### using n_estimators 
n_estimators=[1,2,4,8,16,32,64,100,200]
test_results=[]
accuracy_results=[]
i=0
for estimator in n_estimators:
    
    rf=RandomForestClassifier(n_estimators=estimator,n_jobs=-1)## n_jobs=-1 means the available processors in your system start doing processing on it
    rf.fit(x_train,y_train)
    y_pred=rf.predict(x_test)
    false_positive_rate,true_positive_rate,thresholds=roc_curve(y_test,y_pred)
    roc_auc=auc(false_positive_rate,true_positive_rate)
    test_results.append(roc_auc)
    acc=accuracy_score(y_test,y_pred)
    accuracy_results.append(acc)
    #print(test_results)
    print("auc result at n_estimator ", estimator,"is",test_results[i],)
    print("accuracy result at n_estimator ", estimator,"is",accuracy_results[i],"\n\n")
    i=i+1
    
#print("auc result at different values of n_estimators ",test_results)
#print("accuracy_results at different values of n_estimators",accuracy_results)


    
    

auc result at n_estimator  1 is 0.7421914785265659
accuracy result at n_estimator  1 is 0.7443946188340808 


auc result at n_estimator  2 is 0.7613732812765236
accuracy result at n_estimator  2 is 0.7892376681614349 


auc result at n_estimator  4 is 0.7838227805126463
accuracy result at n_estimator  4 is 0.7982062780269058 


auc result at n_estimator  8 is 0.7662536071974198
accuracy result at n_estimator  8 is 0.7713004484304933 


auc result at n_estimator  16 is 0.7728738754031573
accuracy result at n_estimator  16 is 0.7847533632286996 


auc result at n_estimator  32 is 0.7793668307587847
accuracy result at n_estimator  32 is 0.7847533632286996 


auc result at n_estimator  64 is 0.7851807842471566
accuracy result at n_estimator  64 is 0.7892376681614349 


auc result at n_estimator  100 is 0.788830419283653
accuracy result at n_estimator  100 is 0.7937219730941704 


auc result at n_estimator  200 is 0.7961296893566457
accuracy result at n_estimator  200 is 0.8026905829596412 

### you can see at n_estimator at 200 it gives better performance  with 80% accuracy and auc 0.7961

this way we can tune the hyper parameters to improve the performance